packages/kbn-optimizer/src/new_platform_plugins.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Path from 'path';
+
+import globby from 'globby';
+import loadJsonFile from 'load-json-file';
+
+export interface NewPlatformPlugin {
+  readonly directory: string;
+  readonly id: string;
+  readonly isUiPlugin: boolean;
+}
+
+/**
+ * Helper to find the new platform plugins.
+ */
+export function findNewPlatformPlugins(repoRoot: string) {
+  const manifestPaths = globby.sync(['src/plugins/*/kibana.json', 'x-pack/plugins/*/kibana.json'], {
+    absolute: true,
+    cwd: repoRoot,
+  });
+
+  return manifestPaths.map(path => readNewPlatformPlugin(path));
+}
+
+function readNewPlatformPlugin(manifestPath: string): NewPlatformPlugin {
+  if (!Path.isAbsolute(manifestPath)) {
+    throw new TypeError('expected new platform manifest path to be absolute');
+  }
+
+  const manifest = loadJsonFile.sync(manifestPath);
+  if (!manifest || typeof manifest !== 'object' || Array.isArray(manifest)) {
+    throw new TypeError('expected new platform plugin manifest to be a JSON encoded object');
+  }
+
+  if (typeof manifest.id !== 'string') {
+    throw new TypeError('expected new platform plugin manifest to have a string id');
+  }
+
+  return {
+    directory: Path.dirname(manifestPath),
+    id: manifest.id,
```

Comment:
@joshdover what if we split the current [discover](https://github.com/elastic/kibana/blob/bbf703a0ff46395114947d15481384ad4ddbc071/src/core/server/plugins/discovery/plugins_discovery.ts#L45) implementation into discovery and plugin initialization steps? Then optimizer can re-use discovery functionality.

Reply:
Sounds like a reasonable change, and I don't think it'd be particularly hard at all.